dht_net.go

```diff
@@ -57,36 +58,29 @@ func (dht *IpfsDHT) handleNewMessage(s inet.Stream) {
 	mPeer := s.Conn().RemotePeer()
 
 	for {
-		// receive msg
-		pmes := new(pb.Message)
-		switch err := r.ReadMsg(pmes); err {
+		var pmes pb.Message
+		switch err := r.ReadMsg(&pmes); err {
 		case io.EOF:
-			s.Close()
-			return
-		case nil:
+			return nil
```

Comment:
This will cause us to reset the stream which will throw away any responses that may be queued. This shouldn't break the _current_ go implementation as we always read the response _before_ closing but we don't require that behavior. That is, the following should work:

1. Open a stream.
2. Send a request.
3. Close the stream.
4. Read the response.
5. Read the EOF from the other side.

Reply:
What if s.Close() was retained here? I tried to poke through the code to determine if Close then Reset would signal the equivalent of a "graceful" CancelRead but could not figure it out. The fact that we read a io.EOF means that a Close is sufficient from our end, but do we care? The semantics of Reset are that we're done reading and writing, and this is precisely what occurs at this point in the code.

Reply:
I'm not sure what signal the peer gets if we Reset after they Close. All in all we just don't care, and neither should they.

Reply:
> The semantics of Reset are that we're done reading and writing, and this is precisely what occurs at this point in the code.

The semantics of Reset is, effectively, an abort.

---

> What if s.Close() was retained here?

...

> I'm not sure what signal the peer gets if we Reset after they Close. All in all we just don't care, and neither should they.


That should be fine. A Reset a stream is fully closed (`Close()` + read an `io.EOF) is safe in all of our implementations and, IMO, we should guarantee this. However, we should probably:

1. Document it.
2. Add it to our stream muxer tests.

Unfortunately, I'm pretty sure that a Reset sent _before_ we receive an EOF will cause the other side to receive a Reset instead of an EOF.

Reply:
@Stebalien Let me know if the version is satisfactory.